src/runtime/vdom/update-element.ts

```diff
@@ -23,15 +23,36 @@ export const updateElement = (
 
   if (BUILD.updatable) {
     // remove attributes no longer present on the vnode by setting them to undefined
-    for (memberName in oldVnodeAttrs) {
+    for (memberName of sortedAttrNames(Object.keys(oldVnodeAttrs))) {
       if (!(memberName in newVnodeAttrs)) {
         setAccessor(elm, memberName, oldVnodeAttrs[memberName], undefined, isSvgMode, newVnode.$flags$);
       }
     }
   }
 
   // add new & update changed attributes
-  for (memberName in newVnodeAttrs) {
+  for (memberName of sortedAttrNames(Object.keys(newVnodeAttrs))) {
     setAccessor(elm, memberName, oldVnodeAttrs[memberName], newVnodeAttrs[memberName], isSvgMode, newVnode.$flags$);
   }
 };
+
+/**
+ * Sort a list of attribute names to ensure that all the attribute names which
+ * are _not_ `"ref"` come before `"ref"`. Preserve the order of the non-ref
+ * attributes.
+ *
+ * **Note**: if the supplied attributes do not include `'ref'` then the same
+ * (by reference) array will be returned without modification.
+ *
+ * @param attrNames attribute names to sort
+ * @returns a list of attribute names, sorted if they include `"ref"`
+ */
+function sortedAttrNames(attrNames: string[]): string[] {
+  if (attrNames.includes('ref')) {
+    // we need to sort these to ensure that `'ref'` is the last attr
+    return [...attrNames.filter((attr) => attr !== 'ref'), 'ref'];
+  } else {
+    // no need to sort, return the original array
+    return attrNames;
+  }
```

Comment:
This is totally just a personal preference but wonder what you think about this little tweak:
```suggestion
  return attrNames.includes('ref')
    // we need to sort these to ensure that `'ref'` is the last attr
    ? [...attrNames.filter((attr) => attr !== 'ref'), 'ref']
    // no need to sort, return the original array
    : attrNames;
```

Reply:
sure, why not!